File: package/umt_python/tests/unit/url/test_parse_query_string.py
import unittest

from src.url import parse_query_string


class TestParseQueryString(unittest.TestCase):
    def test_parse_with_leading_question_mark(self):
        result = parse_query_string("?page=1&q=search")
        self.assertEqual(result, {"page": "1", "q": "search"})

    def test_parse_without_leading_question_mark(self):
        result = parse_query_string("foo=bar&baz=qux")
        self.assertEqual(result, {"foo": "bar", "baz": "qux"})

    def test_parse_full_url(self):
        result = parse_query_string("https://example.com?a=1&b=2")
        self.assertEqual(result, {"a": "1", "b": "2"})

    def test_empty_string(self):
        result = parse_query_string("")
        self.assertEqual(result, {})

    def test_question_mark_only(self):
        result = parse_query_string("?")
        self.assertEqual(result, {})

    def test_encoded_values(self):
        result = parse_query_string("?q=hello%20world")
        self.assertEqual(result, {"q": "hello world"})

    def test_single_parameter(self):
        result = parse_query_string("?key=value")
        self.assertEqual(result, {"key": "value"})

    def test_url_with_no_query_string(self):
        result = parse_query_string("https://example.com/path")
        self.assertEqual(result, {})

    def test_reject_proto_key(self):
        result = parse_query_string("?__proto__=polluted&safe=value")
        self.assertEqual(result, {"safe": "value"})
        self.assertNotIn("__proto__", result)

    def test_reject_constructor_and_prototype_keys(self):
        result = parse_query_string("?constructor=bad&prototype=bad&ok=good")
        self.assertEqual(result, {"ok": "good"})
        self.assertNotIn("constructor", result)
        self.assertNotIn("prototype", result)


if __name__ == "__main__":
    unittest.main()

File: package/umt_rust/tests/url/test_parse_query_string.rs
use umt_rust::url::umt_parse_query_string;

#[test]
fn test_parse_query_string_with_leading_question_mark() {
    let result = umt_parse_query_string("?page=1&q=search");
    assert_eq!(result.get("page").unwrap(), "1");
    assert_eq!(result.get("q").unwrap(), "search");
    assert_eq!(result.len(), 2);
}

#[test]
fn test_parse_query_string_without_leading_question_mark() {
    let result = umt_parse_query_string("foo=bar&baz=qux");
    assert_eq!(result.get("foo").unwrap(), "bar");
    assert_eq!(result.get("baz").unwrap(), "qux");
    assert_eq!(result.len(), 2);
}

#[test]
fn test_parse_full_url() {
    let result = umt_parse_query_string("https://example.com?a=1&b=2");
    assert_eq!(result.get("a").unwrap(), "1");
    assert_eq!(result.get("b").unwrap(), "2");
    assert_eq!(result.len(), 2);
}

#[test]
fn test_parse_empty_string() {
    let result = umt_parse_query_string("");
    assert!(result.is_empty());
}

#[test]
fn test_parse_question_mark_only() {
    let result = umt_parse_query_string("?");
    assert!(result.is_empty());
}

#[test]
fn test_parse_encoded_values() {
    let result = umt_parse_query_string("?q=hello%20world");
    assert_eq!(result.get("q").unwrap(), "hello world");
}

#[test]
fn test_parse_single_parameter() {
    let result = umt_parse_query_string("?key=value");
    assert_eq!(result.get("key").unwrap(), "value");
    assert_eq!(result.len(), 1);
}

#[test]
fn test_parse_url_with_no_query_string() {
    let result = umt_parse_query_string("https://example.com/path");
    assert!(result.is_empty());
}

#[test]
fn test_reject_proto_key() {
    let result = umt_parse_query_string("?__proto__=polluted&safe=value");
    assert_eq!(result.get("safe").unwrap(), "value");
    assert!(!result.contains_key("__proto__"));
    assert_eq!(result.len(), 1);
}

#[test]
fn test_reject_constructor_and_prototype_keys() {
    let result = umt_parse_query_string("?constructor=bad&prototype=bad&ok=good");
    assert_eq!(result.get("ok").unwrap(), "good");
    assert!(!result.contains_key("constructor"));
    assert!(!result.contains_key("prototype"));
    assert_eq!(result.len(), 1);
}
